runtime/src/tasks/fifo/tasks-fifo.c

```diff
@@ -1497,3 +1535,105 @@ uint32_t chpl_task_getNumIdleThreads(void) {
   assert(numIdleThreads >= 0);
   return numIdleThreads;
 }
+
+
+//
+// Tasking callback support.
+//
+int chpl_task_install_callback(chpl_task_cb_event_kind_t event_kind,
+                               chpl_task_cb_info_kind_t info_kind,
+                               chpl_task_cb_fn_t cb_fn) {
+  int i;
+
+  if (event_kind >= chpl_task_cb_num_event_kinds) {
+    errno = ERANGE;
+    return -1;
+  }
+
+  i = cb_info[event_kind].count;
+
+  if (i >= MAX_CBs_PER_EVENT) {
+    errno = ENOMEM;
+    return -1;
+  }
+
+  cb_info[event_kind].count++;
+  cb_info[event_kind].fns[i]= cb_fn;
+  cb_info[event_kind].info_kinds[i] = info_kind;
+
+  return 0;
+}
+
+
+int chpl_task_uninstall_callback(chpl_task_cb_event_kind_t event_kind,
+                                 chpl_task_cb_fn_t cb_fn) {
+  int i;
+
+  if (event_kind >= chpl_task_cb_num_event_kinds) {
+    errno = ERANGE;
+    return -1;
+  }
+
+  for (i = 0; i < cb_info[event_kind].count; i++) {
+    if (cb_info[event_kind].fns[i] == cb_fn)
+      break;
+  }
+
+  if (i >= cb_info[event_kind].count) {
+    errno = ENOENT;
+    return -1;
+  }
+
+  for ( ; i < cb_info[event_kind].count - 1; i++) {
+    cb_info[event_kind].fns[i] = cb_info[event_kind].fns[i + 1];
+    cb_info[event_kind].info_kinds[i] = cb_info[event_kind].info_kinds[i + 1];
+  }
+
+  cb_info[event_kind].count--;
+
+  return 0;
+}
+
+
+static void do_callbacks(chpl_task_cb_event_kind_t event_kind,
+                         const task_pool_p ptask) {
+  struct cb_info* cbp;
+
+  assert(event_kind < chpl_task_cb_num_event_kinds);
```

Comment:
The install and uninstall are (or can be) called from outside the runtime, so I'm just doing regular argument edit-checks there. The do-callbacks function is entirely internal to tasks=fifo, so it shouldn't screw up like this. Also, I wanted to make this path fast, because we call it even if there are no callbacks installed. In an optimized runtime the assert will not be present in the emitted code.